packages/@lwc/engine-dom/src/renderer-factory.ts

```diff
@@ -0,0 +1,386 @@
+    }
+
+    let createFragment: (html: string) => Node | null;
+    // IE11 lacks support for this feature
+    const SUPPORTS_TEMPLATE = typeof HTMLTemplateElement === 'function';
+    if (SUPPORTS_TEMPLATE) {
+        // Parse the fragment HTML string into DOM
+        createFragment = function (html: string) {
+            const template = document.createElement('template');
+            template.innerHTML = html;
+            return template.content.firstChild;
+        };
+    } else {
+        // In browsers that don't support <template> (e.g. IE11), we need to be careful to wrap elements like
+        // <td> in the proper container elements (e.g. <tbody>), because otherwise they will be parsed as null.
+
+        // Via https://github.com/webcomponents/polyfills/blob/ee1db33/packages/template/template.js#L273-L280
+        // With other elements added from:
+        // https://github.com/sindresorhus/html-tags/blob/95dcdd5/index.js
+        // Using the test:
+        // document.createRange().createContextualFragment(`<${tag}></${tag}>`).firstChild === null
+        // And omitting <html>, <head>, and <body> as these are not practical in an LWC component.
+        const topLevelWrappingMap: { [key: string]: string[] } = {
+            caption: ['table'],
+            col: ['colgroup', 'table'],
+            colgroup: ['table'],
+            option: ['select'],
+            tbody: ['table'],
+            td: ['tr', 'tbody', 'table'],
+            th: ['tr', 'tbody', 'table'],
+            thead: ['table'],
+            tfoot: ['table'],
+            tr: ['tbody', 'table'],
+        };
+
+        // Via https://github.com/webcomponents/polyfills/blob/ee1db33/packages/template/template.js#L282-L288
+        const getTagName = function (text: string) {
+            return (/<([a-z][^/\0>\x20\t\r\n\f]+)/i.exec(text) || ['', ''])[1].toLowerCase();
+        };
+
+        // Via https://github.com/webcomponents/polyfills/blob/ee1db33/packages/template/template.js#L295-L320
+        createFragment = function (html: string) {
+            const wrapperTags = topLevelWrappingMap[getTagName(html)];
+            if (!isUndefined(wrapperTags)) {
+                for (const wrapperTag of wrapperTags) {
+                    html = `<${wrapperTag}>${html}</${wrapperTag}>`;
+                }
+            }
+
+            // For IE11, the document title must not be undefined, but it can be an empty string
+            // https://developer.mozilla.org/en-US/docs/Web/API/DOMImplementation/createHTMLDocument#browser_compatibility
+            const doc = document.implementation.createHTMLDocument('');
+            doc.body.innerHTML = html;
+
+            let content: Node = doc.body;
+            if (!isUndefined(wrapperTags)) {
+                for (let i = 0; i < wrapperTags.length; i++) {
+                    content = content.firstChild!;
+                }
+            }
+
+            return content.firstChild;
+        };
+    }
```

Comment:
I see it as a downside of this PR that we are now required to move all these functions into one big file. Conceptually, I think it was nice for `createFragment` to live in its own file.

Reply:
Talked to Nolan about this offline, we have some ideas of creating a separate `@lwc/renderer-dom` package where the renderer code can be modularized and using rollout to create the renderer-factory. Will tackle that as a separate follow up PR.

Reply:
That was the original idea... but with the new twist, initially we were saying that locker will depend on that package directly to roll it into lockers source... but with the factory, we can have the best of both worlds.

Reply:
I thought about this some more... I'm not sure we even need separate packages. My main concern is that Locker implementation details don't bleed too much into LWC's source code. If we split the renderer APIs into their own file that exports top-level functions, and if we have a build step in `@lwc/engine-dom`, then I think we can do it all in one package without extra packages needed.

A good example of why we need this is pivots (#2724) – there is no way we want to put all that code inside of a single function with no external dependencies. 🙂